offset_manager.go

```diff
@@ -100,13 +132,81 @@ func (om *offsetManager) ManagePartition(topic string, partition int32) (Partiti
 	}
 
 	if topicManagers[partition] != nil {
-		return nil, ConfigurationError("That topic/partition is already being managed")
+		return nil, ConfigurationError(fmt.Sprintf("topic:%s/partition:%d is already being managed", topic, partition))
 	}
 
 	topicManagers[partition] = pom
 	return pom, nil
 }
 
+func (om *offsetManager) RemovePartitions(topicPartitions map[string][]int32) error {
+	var errs ConsumerErrors
+	var errsLock sync.Mutex
+
+	var wg sync.WaitGroup
+	for topic, partitions := range topicPartitions {
+		for _, partition := range partitions {
+			wg.Add(1)
+			go func(topic string, partition int32) {
+				defer wg.Done()
+
+				om.pomsLock.RLock()
+				pom := om.poms[topic][partition]
+				om.pomsLock.RUnlock()
+				err := pom.Close()
```

Comment:
I met panic during local test.

```go
panic: send on closed channel

goroutine 67 [running]:
github.com/IBM/sarama.(*partitionOffsetManager).handleError(0x1400013a300, {0x104c21ac8, 0x104fb5e30})
    /Users/wwwjfy/Develop/streamsdk/tmp/vendor/github.com/IBM/sarama/offset_manager.go:748 +0xac
github.com/IBM/sarama.(*offsetManager).handleResponse(0x140004961a0, 0x14000494008, 0x140003414a0, 0x1400039bde0)
    /Users/wwwjfy/Develop/streamsdk/tmp/vendor/github.com/IBM/sarama/offset_manager.go:491 +0x338
github.com/IBM/sarama.(*offsetManager).flushToBroker(0x140004961a0)
    /Users/wwwjfy/Develop/streamsdk/tmp/vendor/github.com/IBM/sarama/offset_manager.go:381 +0x80
github.com/IBM/sarama.(*offsetManager).Commit(0x140004961a0)
    /Users/wwwjfy/Develop/streamsdk/tmp/vendor/github.com/IBM/sarama/offset_manager.go:357 +0x20
github.com/IBM/sarama.(*offsetManager).mainLoop(0x140004961a0)
    /Users/wwwjfy/Develop/streamsdk/tmp/vendor/github.com/IBM/sarama/offset_manager.go:341 +0x88
github.com/IBM/sarama.withRecover(0x0?)
    /Users/wwwjfy/Develop/streamsdk/tmp/vendor/github.com/IBM/sarama/utils.go:43 +0x40
created by github.com/IBM/sarama.newOffsetManagerFromClient in goroutine 23
    /Users/wwwjfy/Develop/streamsdk/tmp/vendor/github.com/IBM/sarama/offset_manager.go:113 +0x230
exit status 2
```

It happened during a partition is revoked from this consumer.

I tried to track how this happens, and I think this is the case:

goroutine 1:
1.1. consumer is informed a partition is revoked
1.2. it comes here to do cleanup, and this line leads to `errors` channel closure at line 755
1.3. later this pom is removed from `om.poms` at line 180

goroutine 2:

2.1 in the `mainLoop`, `om.Commit` is triggered every second by default
2.2 it constructs requests based on `om.poms` (line 440)
2.3 it gets response, and if it's an error `ErrIllegalGeneration`, send to `pom.errors` (line 490)

In a race condition, 2.3 happens between 1.2 and 1.3, a panic will happen.

Reply:
propose change:

```diff
iff --git offset_manager.go offset_manager.go
index 6c40a88..4cfe3f5 100644
--- offset_manager.go
+++ offset_manager.go
@@ -150,10 +150,15 @@ func (om *offsetManager) RemovePartitions(topicPartitions map[string][]int32) er
                        go func(topic string, partition int32) {
                                defer wg.Done()
 
-                               om.pomsLock.RLock()
+                               om.pomsLock.Lock()
                                pom := om.poms[topic][partition]
-                               om.pomsLock.RUnlock()
                                err := pom.Close()
+                               delete(om.poms[topic], partition)
+                               if len(om.poms[topic]) == 0 {
+                                       delete(om.poms, topic)
+                               }
+
+                               om.pomsLock.Unlock()
                                if err != nil {
                                        errsLock.Lock()
                                        var consumerErrs ConsumerErrors
@@ -174,17 +179,6 @@ func (om *offsetManager) RemovePartitions(topicPartitions map[string][]int32) er
                }
        }
 
-       om.pomsLock.Lock()
-       for topic, partitions := range topicPartitions {
-               for _, partition := range partitions {
-                       delete(om.poms[topic], partition)
-                       if len(om.poms[topic]) == 0 {
-                               delete(om.poms, topic)
-                       }
-               }
-       }
-
-       om.pomsLock.Unlock()
        if len(errs) > 0 {
                return errs
        }
```

Reply:
Thanks for troubleshooting. I think this propose change works, and I also move the logic of last offset flush ahead.